ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
+  final Histogram responseProcessingWaitTimeInMs;
+  final Histogram responseProcessingTimeInMs;
+  final Histogram roundTripTimeInMs;
+
+  final Meter operationRate;
+  final Meter operationErrorRate;
+
+  /**
+   * Instantiates {@link CryptoJobMetrics} for {@code requestType} and attaches all the metrics related to the
+   * cryptoJob to the given {@code ownerClass}. The metrics are also registered in the provided {@code metricRegistry}.
+   * @param ownerClass the {@link Class} that is supposed to own the metrics created by this tracker.
+   * @param requestType the type of request for which a tracker is being created.
+   * @param metricRegistry the {@link MetricRegistry} to use to register the created metrics.
+   */
+  CryptoJobMetrics(Class ownerClass, String requestType, MetricRegistry metricRegistry) {
+    requestQueuingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + REQUEST_QUEUING_TIME_SUFFIX));
+    requestProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + REQUEST_PROCESSING_TIME_SUFFIX));
+    responseProcessingWaitTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_WAIT_TIME_SUFFIX));
+    responseProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_TIME_SUFFIX));
+    roundTripTimeInMs = metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + ROUND_TRIP_TIME_SUFFIX));
+
+    operationRate = metricRegistry.meter(MetricRegistry.name(ownerClass, requestType + OPERATION_RATE_SUFFIX));
+    operationErrorRate = metricRegistry.meter(
+        MetricRegistry.name(ownerClass, requestType + OPERATION_ERROR_SUFFIX + OPERATION_RATE_SUFFIX));
+  }
+}
+
+/**
+ * Construct to support end-to-end metrics tracking for crypto jobs. Usually accompanies a single
+ * {@link CryptoJob} i.e. there is a one-to-one mapping b/w a {@link CryptoJob} and a CryptoJobMetricsTracker
+ * instance.
+ * <p/>
+ * A brief description of how the tracker works :-
+ * - When an object of type {@link CryptoJob} is instantiated, it is also expected to be associated with a
+ *    unique instance of CryptoJobMetricsTracker (tracker). The Tracker will be instantiated either with
+ *    EncryptJobMetrics or DecryptJobMetrics
+ * - As the cryptoJob passed through various phases, metrics associated with these are updated via methods exposed
+ *    by this tracker.
+ */
+class CryptoJobMetricsTracker {
+  private final CryptoJobMetrics cryptoJobMetrics;
+  private long trackerStartTimeMs;
+  private long lastSeenTimeMs;
+
+  CryptoJobMetricsTracker(CryptoJobMetrics cryptoJobMetrics) {
+    this.cryptoJobMetrics = cryptoJobMetrics;
+  }
+
+  /**
+   * Starts the operation tracker for the corresponding CryptoJob
+   */
+  void startTracker() {
+    trackerStartTimeMs = System.currentTimeMillis();
+    lastSeenTimeMs = trackerStartTimeMs;
```

Comment:
why is this set here?

Reply:
what do you mean by set here. Sorry I don't get you. Do you suggest it to be set elsewhere?